File: python/scomlVarSocket.py
import socket


def get_max_num(num):
    """
    描述: 根据十进制位数计算其表示最大数值
    :param num: 输入十进制位数
    :return: 最大数值
    """
    if not isinstance(num, int):
        raise RuntimeError("num should be int.")
    if num <= 0:
        raise RuntimeError("num should exceed 0.")
    str_num = ''
    for i in range(num):
        str_num += '9'
    return int(str_num)


def scoml_var_send(handle_socket, msg, ori_len=12):
    """
    描述: 可变长度 socket 接收函数
    :param handle_socket: 指用哪个 socket 来发送数据
    :param msg: bytes, or bytes from str with encoding ascii
    :param ori_len: default 12
    :return: 0
    """
    # if not (isinstance(msg, str) or isinstance(msg, bytes)):
    #     raise RuntimeError("msg data type must be str or bytes")
    # elif isinstance(msg, str):
    #     msg = bytes(msg, 'utf-8')
    if not isinstance(msg, bytes):
        raise RuntimeError("msg data type must be bytes")
    if len(msg) > get_max_num(ori_len):
        raise RuntimeError("msg data exceeds the length which ori_len can describes. "
                           "You should increase the ori_len parameter in both server and client.")
    ori_sent = 0
    ori_msg = bytes(str(len(msg)).zfill(ori_len), 'ascii')
    while ori_sent < ori_len:
        sent = handle_socket.send(ori_msg[ori_sent:])
        if sent == 0:
            raise RuntimeError("socket connection broken")
        ori_sent += sent
    msg_sent = 0
    while msg_sent < len(msg):
        sent = handle_socket.send(msg[msg_sent:])
        if sent == 0:
            raise RuntimeError("socket connection broken")
        msg_sent += sent
    return 0


def scoml_var_receive(handle_socket, ori_len=12):
    """
    描述: 可变长度 socket 发送函数 
    :param handle_socket: 指用哪个 socket 来接收数据
    :param ori_len: 表示数据体长度的数位, 默认值 12
    :return: bytes
    """
    ori_chunks = []
    ori_recd = 0
    while ori_recd < ori_len:
        recd = handle_socket.recv(min(ori_len-ori_recd, 2048))
        if recd == b'':
            raise RuntimeError("socket connection broken")
        ori_chunks.append(recd)
        ori_recd += len(recd)
    msg_len = int(b''.join(ori_chunks).decode('ascii'))

    chunks = []
    msg_recd = 0
    while msg_recd < msg_len:
        recd = handle_socket.recv(min(msg_len - msg_recd, 2048))
        if recd == b'':
            raise RuntimeError("socket connection broken")
        chunks.append(recd)
        msg_recd += len(recd)
        if msg_recd > get_max_num(ori_len):
            raise RuntimeError("received data exceeds the length which ori_len can describes. "
                               "You should increase the ori_len parameter in both server and client.")
    return b''.join(chunks)


class ScomlServerSocket:
    """
    描述: 服务端 socket 类
    """
    def __init__(self):
        pass

    def response(self, msg):
        """
        execute function
        :param msg: received msg of bytes type
        :return: response msg of bytes type
        """
        raise NotImplementedError("You should write your execute code here.")

    def start(self, port, ori_len=12):
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.bind(("localhost", port))
        sock.listen()
        print("server running at port "+str(port))
        while True:
            try:
                handle_socket, addr = sock.accept()
                message = scoml_var_receive(handle_socket, ori_len)
                answer = self.response(message)
                if answer is None:
                    continue
                if not isinstance(answer, bytes):
                    raise RuntimeError("answer is not bytes.")
                scoml_var_send(handle_socket, answer, ori_len)
                handle_socket.close()
            except Exception:
                print("error when response.")
                sock.close()
                return


def scoml_var_request(ip, port, msg, ori_len=12):
    """

    :param ip: str
    :param port:
    :param msg: bytes, or bytes from str with encoding utf-8
    :param ori_len:
    :return:
    """
    if not isinstance(msg, bytes):
        return "msg data type must be bytes"
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((ip, port))
    scoml_var_send(sock, msg)
    answer = scoml_var_receive(sock, ori_len)
    sock.close()
    return answer

File: cpp/scomlVarSocket.cpp
#include <sstream>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <sys/wait.h>
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

namespace scoml{
// get sockaddr, IPv4 or IPv6:
void *get_in_addr(struct sockaddr *sa){
    if (sa->sa_family == AF_INET) {
        return &(((struct sockaddr_in*)sa)->sin_addr);
    }
    return &(((struct sockaddr_in6*)sa)->sin6_addr);
}

long long getMaxNum(int num){
    std::string strNum;
    for(int i=0; i<num; ++i)
        strNum += "9";
    return std::stoll(strNum);
}

/****************************************************************
 * 描述: 可变长度 socket 接收函数
 * 输入参数:
 *      handleSocket: socket descript file
 *      msg: 要发送的字符
 *      oriLen; 表达数据字节长度的文本字符数
 * 输出参数:
 *      int: 表示成功与否
****************************************************************/
int varSend(int handleSocket, std::string &msg, int oriLen=12){
    long long msgLen = msg.size();
    if(msgLen > getMaxNum(oriLen)){
        std::cout << "msg data exceeds the length which ori_len can describes. You should increase the ori_len parameter in both server and client." << std::endl;
        return -1;
    }
    std::string tmp = std::to_string(msgLen);
    std::string zeroString(oriLen-tmp.size(), '0');
    const char* oriChar = (zeroString + tmp).c_str();
    int oriSent = 0;
    while (oriSent < oriLen){
        int sendNum = send(handleSocket, oriChar+oriSent, std::min(oriLen - oriSent, 2048), 0);
        if(sendNum == -1){
            std::cout << "socket connection broken." << std::endl;
            return -1;
        }
        oriSent += sendNum;
    }
    long long msgSent = 0;
    const char* msgChar = msg.c_str();
    while (msgSent < msgLen){
        long long sendNum = send(handleSocket, msgChar+msgSent, std::min(msgLen - msgSent, (long long)2048), 0);
        if(sendNum == -1){
            std::cout << "socket connection broken." << std::endl;
            return -1;
        }
        msgSent += sendNum;
    }
    return 0;
}

/****************************************************************
 * 描述: 可变长度 socket 发送函数
 * 输入参数:
 *      handleSocket: socket descript file
 *      oriLen; 表达数据字节长度的文本字符数
 *      msg: 接收数据变量
 * 输出参数:
 *      int: 表示成功与否
****************************************************************/
int varRecv(int handleSocket, std::string &msg, int oriLen=12){
    char *oriChar = new char[oriLen];
    int oriRecd = 0;
    while(oriRecd < oriLen){
        int recvNum = recv(handleSocket, oriChar+oriRecd, std::min(oriLen - oriRecd, 2048), 0);
        if(recvNum == -1){
            std::cout << "socket connection broken." << std::endl;
            return -1;
        }
        oriRecd += recvNum; 
    }
    std::string oriString(oriLen, '0');
    char *pOriChar = oriChar;
    for(int i=0; i<oriLen; ++i){
        oriString[i] = *pOriChar++;
    }
    delete oriChar;
    // msg
    std::istringstream oriStream(oriString);
    long long msgLength; oriStream >> msgLength;
    char *msgChar = new char[msgLength];
    long long msgRecd = 0;
    while(msgRecd < msgLength){
        long long recvNum = recv(handleSocket, msgChar+msgRecd, std::min(msgLength- msgRecd, (long long)2048), 0);
        if(recvNum == -1){
            std::cout << "socket connection broken." << std::endl;
            return -1;
        }
        msgRecd += recvNum; 
        if(msgRecd > getMaxNum(oriLen)){
            std::cout << "msg data exceeds the length which ori_len can describes. You should increase the ori_len parameter in both server and client." << std::endl;
            return -1;
        }
    }
    std::string msgString(msgLength, '0');
    char *pMsgChar = msgChar;
    for(long long i=0; i<msgLength; ++i){
        msgString[i] = *pMsgChar++;
    }
    delete msgChar;
    // msg = msgString;
    std::swap(msg, msgString);
    return 0;
}


std::string var_request(std::string ip, std::string port, std::string msg, int oriLen=12){
    struct addrinfo hints, *p;
    char addrPres[INET6_ADDRSTRLEN];
    int sockfd;
    memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;
    if ((getaddrinfo(ip.c_str(), port.c_str(), &hints, &p)) != 0){
        std::cout << "getaddrinfo failed." << std::endl;
        exit(-1);
    }
    if (p == NULL){
        std::cout << "cann't getaddrinfo." << std::endl;
        exit(-1);
    }
    if ((sockfd = socket(p->ai_family, p->ai_socktype, p->ai_protocol)) == -1){
        std::cout << "create client socket failed." << std::endl;
        exit(-1);
    }
    if (connect(sockfd, p->ai_addr, p->ai_addrlen) == -1){
        std::cout << "client connect error." << std::endl;
        exit(-1);
    }
    inet_ntop(p->ai_family, get_in_addr((struct sockaddr *)p->ai_addr), addrPres, sizeof(addrPres));
    std::cout << "client: connecting to " << addrPres << std::endl;
    freeaddrinfo(p);
    // msg
    if(varSend(sockfd, msg, oriLen)!=0){
        std::cout << "send error." << std::endl;
        exit(-1);
    }
    std::string answer;
    if(varRecv(sockfd, answer, oriLen)!=0){
        std::cout << "recv error." << std::endl;
        exit(-1);
    }
    return answer;
}

} // namespace end
